src/Microsoft.Azure.WebJobs.Host/Tables/TableExtension.cs

```diff
@@ -0,0 +1,371 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Host.Bindings;
+using Microsoft.Azure.WebJobs.Host.Config;
+using Microsoft.Azure.WebJobs.Host.Executors;
+using Microsoft.Azure.WebJobs.Host.Protocols;
+using Microsoft.Azure.WebJobs.Host.Storage;
+using Microsoft.Azure.WebJobs.Host.Storage.Table;
+using Microsoft.WindowsAzure.Storage.Table;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.WebJobs.Host.Tables
+{
+    // Extension for adding Azure Table  bindings. 
+    internal class TableExtension : IExtensionConfigProvider
+    {
+        private IStorageAccountProvider _accountProvider;
+
+        // Property names on TableAttribute
+        private const string RowKeyProperty = "RowKey";
+        private const string PartitionKeyProperty = "PartitionKey";
+
+        public void Initialize(ExtensionConfigContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+            var config = context.Config;
+            _accountProvider = config.GetService<IStorageAccountProvider>();
+            var nameResolver = config.NameResolver;
+
+            // rules for single entity. 
+            var original = new TableAttributeBindingProvider(nameResolver, _accountProvider);
+
+            var builder = new JObjectBuilder(this);
+
+            context.AddBindingRule<TableAttribute>()
+                .AddValidator(ValidateAttribute)
+                .AddConverter<JObject, ITableEntity>(JObjectToTableEntityConverterFunc)
+                .AddConverter<object, ITableEntity>(typeof(ObjectToITableEntityConverter<>))
+                .AddConverter<IStorageTable, IQueryable<OpenType>>(typeof(TableToIQueryableConverter<>))                
+                .WhenIsNull(RowKeyProperty)
+                    .SetPostResolveHook(ToParameterDescriptorForCollector)
+                    .BindToInput<CloudTable>(builder)                    
+                .WhenIsNull(RowKeyProperty)
+                    .SetPostResolveHook(ToParameterDescriptorForCollector)
+                    .BindToInput<IStorageTable>(builder)                    
+                .BindToCollector<ITableEntity>(BuildFromTableAttribute)
+                .WhenIsNotNull(PartitionKeyProperty).WhenIsNotNull(RowKeyProperty)
+                    .BindToInput<JObject>(builder)
+                .BindToInput<JArray>(builder)
+                .Bind(original);
```

Comment:
How strict is the ordering here? Can I get myself into trouble if I call BindToInput().WhenIsNull()? What kinds of errors do I get?